### PR TITLE
Use only discrete-basis translations in `GateDirection`

### LINKED ISSUE #10785: GateDirection pass goes outside allowed basis gates
### Environment

- **Qiskit Terra version**: 0.25
- **Python version**: 3.9
- **Operating system**: MacOS

### What is happening?

Trying to run a clifford simulation on a fake backend with ECR basis gates leads to transpilation errors with unsupported basis gates

### How can we reproduce the issue?

```python
from qiskit import QuantumCircuit, transpile
from qiskit_aer import AerSimulator
from qiskit.providers.fake_provider import FakeSherbrooke, FakeWashington

# Stabilizer simulator for FakeSherbrooke
clifford_basis_gates = ['ecr', 'sx', 'x', 'id', 'z', 's', 'sdg']
backend = AerSimulator.from_backend(
    FakeSherbrooke(),
    method="stabilizer",
    basis_gates=clifford_basis_gates, 
)

# Circuit
qc = QuantumCircuit(2)
qc.h(0)
qc.cx(0, 1)

# Transpile
transpile(qc, backend)
```
Raises
```
TranspilerError: "Unable to translate the operations in the circuit: ['ry', 'h', 'sdg', 'sx', 's', 'ecr', 'x'] to the backend's (or manually specified) target basis: ['break_loop', 'quantum_channel', 's', 'save_expval_var', 'for_loop', 'roerror', 'continue_loop', 'save_probabilities', 'z', 'qerror_loc', 'reset', 'save_clifford', 'snapshot', 'if_else', 'id', 'save_probabilities_dict', 'ecr', 'save_state', 'save_amplitudes_sq', 'sdg', 'save_stabilizer', 'delay', 'while_loop', 'save_expval', 'set_stabilizer', 'sx', 'measure', 'barrier', 'x']. This likely means the target basis is not universal or there are additional equivalence rules needed in the EquivalenceLibrary being used. For more details on this error see: https://qiskit.org/documentation/stubs/qiskit.transpiler.passes.BasisTranslator.html#translation_errors"
```

### What should happen?

Running with a manually specified basis gates works (though the output is not at all optimized in terms of number of 1Q gates):

```python
transpile(qc, basis_gates=["ecr", "x", "sx", "id", "z", "s", "sdg"])
```
```
global phase: 3π/2
     ┌───┐┌────┐┌───┐ ┌─────┐     ┌──────┐┌───┐
q_0: ┤ S ├┤ √X ├┤ S ├─┤ Sdg ├─────┤0     ├┤ X ├
     ├───┤├───┬┘├───┴┐└┬───┬┘┌───┐│  Ecr │└───┘
q_1: ┤ S ├┤ S ├─┤ √X ├─┤ S ├─┤ S ├┤1     ├─────
     └───┘└───┘ └────┘ └───┘ └───┘└──────┘     
```

### Any suggestions?

Running transpilation with a callback looks like the issues is due to the `GateDirection` pass which inserts non-Clifford gates outside the allowed basis gates into the circuit and causes the rest of the transpilation to fail. 

```python
def callback(**kwargs):
    print(f"{type(kwargs['pass_']).__name__} output gates: {kwargs['dag'].count_ops()}")

transpile(qc, backend, callback=callback).draw()
```

```
ContainsInstruction, gates: {'h': 1, 'cx': 1}
UnitarySynthesis, gates: {'h': 1, 'cx': 1}
HighLevelSynthesis, gates: {'h': 1, 'cx': 1}
Unroll3qOrMore, gates: {'h': 1, 'cx': 1}
SetLayout, gates: {'h': 1, 'cx': 1}
TrivialLayout, gates: {'h': 1, 'cx': 1}
CheckMap, gates: {'h': 1, 'cx': 1}
FullAncillaAllocation, gates: {'h': 1, 'cx': 1}
EnlargeWithAncilla, gates: {'h': 1, 'cx': 1}
ApplyLayout, gates: {'h': 1, 'cx': 1}
CheckMap, gates: {'h': 1, 'cx': 1}
UnitarySynthesis, gates: {'h': 1, 'cx': 1}
HighLevelSynthesis, gates: {'h': 1, 'cx': 1}
UnrollCustomDefinitions, gates: {'h': 1, 'cx': 1}
BasisTranslator, gates: {'s': 6, 'sx': 2, 'sdg': 1, 'ecr': 1, 'x': 1}
CheckGateDirection, gates: {'s': 6, 'sx': 2, 'sdg': 1, 'ecr': 1, 'x': 1}
GateDirection, gates: {'s': 6, 'sx': 2, 'sdg': 1, 'x': 1, 'ry': 2, 'ecr': 1, 'h': 2}   # NON BASIS RY INSERTED HERE
RemoveResetInZeroState, gates: {'s': 6, 'sx': 2, 'sdg': 1, 'x': 1, 'ry': 2, 'ecr': 1, 'h': 2}
Depth, gates: {'s': 6, 'sx': 2, 'sdg': 1, 'x': 1, 'ry': 2, 'ecr': 1, 'h': 2}
FixedPoint, gates: {'s': 6, 'sx': 2, 'sdg': 1, 'x': 1, 'ry': 2, 'ecr': 1, 'h': 2}
Size, gates: {'s': 6, 'sx': 2, 'sdg': 1, 'x': 1, 'ry': 2, 'ecr': 1, 'h': 2}
FixedPoint, gates: {'s': 6, 'sx': 2, 'sdg': 1, 'x': 1, 'ry': 2, 'ecr': 1, 'h': 2}
Optimize1qGatesDecomposition, gates: {'s': 6, 'sx': 2, 'sdg': 1, 'x': 1, 'ry': 2, 'ecr': 1, 'h': 2}
CXCancellation, gates: {'s': 6, 'sx': 2, 'sdg': 1, 'x': 1, 'ry': 2, 'ecr': 1, 'h': 2}
GatesInBasis, gates: {'s': 6, 'sx': 2, 'sdg': 1, 'x': 1, 'ry': 2, 'ecr': 1, 'h': 2}
UnitarySynthesis, gates: {'s': 6, 'sx': 2, 'sdg': 1, 'x': 1, 'ry': 2, 'ecr': 1, 'h': 2}
HighLevelSynthesis, gates: {'s': 6, 'sx': 2, 'sdg': 1, 'x': 1, 'ry': 2, 'ecr': 1, 'h': 2}
UnrollCustomDefinitions, gates: {'s': 6, 'sx': 2, 'sdg': 1, 'x': 1, 'ry': 2, 'ecr': 1, 'h': 2}
# raises exception here
```

This pass should only insert gates supported by the backend, and be updated to work with Clifford basis backends to enable simulation testing

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

This makes it a little more likely that it'll be possible to use the transpiler when targetting a discrete basis, such as a Clifford simulator.  We still in general do not offer full support for discrete-basis translation, and certainly not for optimal discrete-basis transpilation, but this is a relatively easy change that makes us marginally more reliable with them.

### Details and comments

Fix #10785.

I'm vaguely concerned about a potential performance regression here by us making the basis translator and resynthesis routines do more work for real-world backends, but I couldn't immediately think what benchmarks, if any, we have to run about that.

*edit*: I'm not 100% convinced that this PR is something we should do, but it's a proof of principle.

### CODE REVIEW

**Comment by jakelishman** [comment]
The tests here will actually have some problems because they'd need a global phase within control-flow blocks.  It's not actually clear to me how much we physically need to track that (once there's control flow, it's not possible to get a unitary representation of the circuit anyway), but even if we were to, we currently don't have a way to do that via the control-flow builder interface.

**Comment by 1ucian0** [comment]
That means this PR is blocked on https://github.com/Qiskit/qiskit/pull/10800?

**Comment by jakelishman** [comment]
Yeah, or something like that. It's also not 100% clear that this PR is definitely the right direction in the general case - I might need to benchmark to see if there's any performance regression, and if so, make the pass dynamically choose which ecr template to use based on the target.

**Comment by kdk** [comment]
> I'm vaguely concerned about a potential performance regression here by us making the basis translator and resynthesis routines do more work for real-world backends, ...

For the majority of active devices (maybe all that use `ecr` at least), wouldn't we already need to translate and re-synthesize away from `ry`? Is there somewhere this would lead to more effort on the part of the transpiler?

> That means this PR is blocked on https://github.com/Qiskit/qiskit/pull/10800?

Agree that's a bug, but it seems to be pre-existing. Does it need to block the PR here?

> ... make the pass dynamically choose which ecr template to use based on the target.

Agree this seems like the right long term direction.

**Comment by jakelishman** [comment]
> [worried about resynthesis routines doing more work]

The trick here is the existing template inserts only two `ry` gates that need resynthesis for most ecr-supporting backends, whereas the new template inserts six gates of which the four `s` and `sdg`s will need resynthising, and the resulting 1q gate chain will be longer, which means the subsequent 1q optimiser will have to do more work to see if it can combine them.  I don't know that that'll have a measurable performance impact or anything, I was just flagging it as something to check.

> [Does global-phase in the control-flow builders need to block this?]

In practice it's hard and fiddly to update the tests in this PR to respect the new behaviour without it.  Not impossible, but assuming the global-phase PR is not especially contentious, it's much more convenient just to merge that first.

**Comment by jakelishman** [comment]
Since #10800 is merged, I've updated the tests in this PR.

I ran a simple test for any performance regressions with this script:
```python
from qiskit import transpile
from qiskit.circuit.library import EfficientSU2
from qiskit.providers.fake_provider import FakeSherbrooke

qc = EfficientSU2(100, reps=10)
backend = FakeSherbrooke()
transpile(qc, backend=backend, seed_transpiler=0)

%timeit [transpile(qc, backend=backend, seed_transpiler=i) for i in range(10)]
```
since O1 is the level we'd expect to see the largest different at anyway, and Sherbrooke is an ECR-based backend.

On `main` I found it took 38.7(3)s, and this branch took 38.9(1)s, so there's nothing really to mention between the two.

File: qiskit/transpiler/passes/utils/gate_direction.py
Comment (mtreinish) [inline qiskit/transpiler/passes/utils/gate_direction.py:110]: From a performance PoV there isn't a huge difference either way. For each ECR gate we flip we end up doing two extra products here: https://github.com/Qiskit/qiskit/blob/main/qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py#L197 but they're all 2x2 so it'll be fast. There might actually be an immeasurable speed up because we don't have to build 2 matrices for the ry gates when `to_matrix()` is called. From a memory overhead perspective I think with #10314 this is actually more efficient because 2 Ry(..)  is about 1kb and this ends up being 6 pointers (assuming `SGate()` and `SdgGate()` are instantiated somewhere already, `SXGate` should be because it's always in the default IBM basis with ECR).

**Review by mtreinish — APPROVED** [review]
I'm fine with doing this, in the general case it won't fundamentally change the performance or output quality afaict. There might be some subtle changes/edge cases for the output for level 0 in some cases because we run gate direction followed by translation only. So it's using the equivalence library to translate from `s, sdg, sx` to -> target instead of from `ry`, which might return a different transform based on the exact rules in the equivalence library (the other opt levels run `Optimize1qGatesDecomposition` which will normalize to the same sequence based on the target and matrix of the sequence).

Longer term it'd be good if we could handle the translation of `Ry(pi/2)`  to a discrete basis directly in the equivalence library since I don't think anything in there precludes us from adding it in theory (there are probably issues around handling discrete parameter values on matching, but we can address those). But in the short term I think this is fine.

File: qiskit/transpiler/passes/utils/gate_direction.py
Comment (jakelishman) [reply qiskit/transpiler/passes/utils/gate_direction.py:110]: oh, that's actually true about memory usage, the gate arrays and the singleton objects - I hadn't been thinking about that at all.  I think I wrote this PR before we landed any of the singleton-gate stuff, so it wasn't on my mind at the time.

**Pull request merged**
